fix: disable background digests on Windows

emillon opened the pull request:
Fixes https://github.com/ocaml/dune/issues/8228

In some cases when the shared cache is enabled, some temporary directories cannot be removed.

emillon commented:
We can probably come up with a better fix but this one would mitigate the issue.

nojb commented:
This fix is not great; it will leave temporary directories and files laying around that will never be cleaned up. Do you understand when is the `rm_rf` called? Could the issue be solved by delaying this call somehow until other threads are done? For example, in the `Temp` module of `stdune` we register an exit hook to clean all temporary files and directories; could that be relevant here? (I haven't looked at the code, so maybe this is precisely the context in which the bug is triggered...)

Alizter commented:
I don't understand how we run into this issue in the first place. When creating temporary directories in Windows, it is my understanding that we ask windows for a unique temp directory name which gets put into the common temp dir location. Then we can do whatever we like with the files inside and finally delete the temp directory since the name is unqiue.

It appears in this case we are placing a non-unique directory name (from the caching stuff?) directly ourselves into the common windows temp dirs and then getting race conditions when two processes touch the same name. Wouldn't the solution be to fix how we create and delete temp directories in windows rather than ignoring the errors?

rgrinberg commented:
How about we just disable background digests on windows until we understand the issue better?

emillon commented:
> How about we just disable background digests on windows until we understand the issue better?

Sure, that's an even better workaround. I'll change the PR to do that and add that to 3.9.2.

emillon committed:
fix: disable background digests on Windows

Fixes #8228

In some cases when the shared cache is enabled, some temporary
directories cannot be removed.

Signed-off-by: Etienne Millon <me@emillon.org>

emillon commented:
done

rgrinberg commented on src/dune_config/config.ml:
```diff
@@ -116,7 +116,10 @@ let background_digests =
-    ; value = background_default
+    ; value =
+        (match Platform.OS.value with
+        | Linux -> `Enabled
```
Why did you disable it on Mac and the rest of the unixes?

rgrinberg replied:
Oh, I remember that we have that fork bug as well.

emillon replied:
I can introduce a `Toggle.and_`  to express that as `background_default and not windows` if you think it's more explicit.

emillon committed:
Merge branch 'main' into fix-8228
